### PR TITLE
feat: move file patterns to a global level to be able to use it on any analyzer

### PR DESCRIPTION
**Pull request by jerbob92**
## Description

This is the functionality of https://github.com/aquasecurity/fanal/pull/372 moved into Trivy since Fanal was merged into it.

This PR moves the file patterns feature from the config analyzers to a global level so that we can use it on any analyzer.
In favor of https://github.com/aquasecurity/fanal/pull/370, https://github.com/aquasecurity/fanal/pull/357, https://github.com/aquasecurity/fanal/pull/355 as this feature will allow you to set the needed file patterns yourself.

This also makes the config analyzers more like the other analyzers.

This can be enabled in Trivy without changing anything to the config scanner. The other commands will just get a --filter-patterns option.

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [x] I've added tests that prove my fix is effective or that my feature works.
- [x] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [x] I've added usage information (if the PR introduces new options)
- [x] I've included a "before" and "after" example to the description (if the PR is a user interface change).

### CODE REVIEW

**Comment by jerbob92** [comment]
@owenrumney @liamg @knqyf263 
Is anyone going to look at this?

**Comment by jerbob92** [comment]
@owenrumney @liamg @knqyf263 Are you guys straight up collectively ignoring me or is something else going on?

**Comment by liamg** [comment]
Hi @jerbob92 - sorry about this! Entirely my fault. It's been a busy couple of weeks and I totally missed this but that's no excuse, I apologise. Looking now :eyes:

**Comment by jerbob92** [comment]
> Hi @jerbob92 - sorry about this! Entirely my fault. It's been a busy couple of weeks and I totally missed this but that's no excuse, I apologise. Looking now eyes

Thanks! Just wondering whether I have to keep updating this MR :sweat_smile:

**Review by liamg — APPROVED** [review]
LGTM! :rocket: 

Sorry for the delay :snail:

**Comment by liamg** [comment]
If you could  check the conflicts I'll make sure I merge this today :)

**Comment by jerbob92** [comment]
@liamg the current failure is this:
```
    docker_engine_test.go:275: 
        	Error Trace:	/home/runner/work/trivy/trivy/integration/docker_engine_test.go:275
        	Error:      	Received unexpected error:
        	            	Error: No such image: ghcr.io/aquasecurity/trivy-test-images:amazon-2
        	Test:       	TestDockerEngine/amazonlinux:2
        	Messages:   	amazonlinux:2
```

That doesn't seem related to my changes, do you have any idea?

**Comment by liamg** [comment]
@jerbob92 I haven't seen this failure before - I'll raise it internally and see if I can find a solution.

**Comment by liamg** [comment]
Apparently this is a fairly common issue with GHCR being unreliable :grimacing: - I've kicked off the integration tests again :crossed_fingers:

**Review by liamg — APPROVED** [review]

**Comment by knqyf263** [comment]
@liamg Thanks for reviewing.
@jerbob92 I hope to take a look shortly.

File: pkg/fanal/analyzer/analyzer.go [hunk 1]
```diff
@@ -313,14 +339,24 @@ func (ag AnalyzerGroup) AnalyzeFile(ctx context.Context, wg *sync.WaitGroup, lim
...
 
-		// filepath extracted from tar file doesn't have the prefix "/"
-		if !a.Required(strings.TrimLeft(filePath, "/"), info) {
+		filePatternMatch := false
+		for _, pattern := range ag.filePatterns[a.Type()] {
+			if pattern.MatchString(cleanPath) {
+				filePatternMatch = true
+				break
+			}
+		}
```
Comment (knqyf263) [inline pkg/fanal/analyzer/analyzer.go:357]: Having it in a dedicated method seems better.

func (ag AnalyzerGroup) filePatternMatch(analyzerType Type, filePath string) {
	for _, pattern := range ag.filePatterns[analyzerType] {
		if pattern.MatchString(cleanPath) {
			return true
		}
	}
	return false
}

**Review by knqyf263 — COMMENTED** [review]
Could you also update this doc?
https://github.com/aquasecurity/trivy/blob/2580ea1583265d669495d8d5bebca86117057c11/docs/docs/references/customization/config-file.md?plain=1#L198-L200

**Comment by jerbob92** [comment]
> Could you also update this doc?
> 
> https://github.com/aquasecurity/trivy/blob/2580ea1583265d669495d8d5bebca86117057c11/docs/docs/references/customization/config-file.md?plain=1#L198-L200

Moving it to scan options?

**Comment by jerbob92** [comment]
> [CLA assistant check](https://cla-assistant.io/aquasecurity/trivy?pullRequest=2539) Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/aquasecurity/trivy?pullRequest=2539) before we can accept your contribution.You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/aquasecurity/trivy?pullRequest=2539) it.

Huh? Didn't I already sign this?

**Comment by afdesk** [comment]
> > [CLA assistant check](https://cla-assistant.io/aquasecurity/trivy?pullRequest=2539) Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/aquasecurity/trivy?pullRequest=2539) before we can accept your contribution.You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/aquasecurity/trivy?pullRequest=2539) it.
> 
> Huh? Didn't I already sign this?

@jerbob92 it seems CLA was updated

**Review by knqyf263 — APPROVED** [review]

**Comment by knqyf263** [comment]
Thanks for your contribution and patience🙇

**Pull request merged**